ssl/quic/quic_impl.c

```diff
@@ -4868,21 +4868,26 @@ int ossl_quic_conn_poll_events(SSL *ssl, uint64_t events, int do_tick,
     return 1;
 }
 
+QUIC_TAKES_LOCK
 int ossl_quic_get_notifier_fd(SSL *ssl)
 {
     QCTX ctx;
     QUIC_REACTOR *rtor;
     RIO_NOTIFIER *nfy;
+    int nfd;
 
     if (!expect_quic_any(ssl, &ctx))
         return -1;
 
+    qctx_lock(&ctx);
```

Comment:
Something seems off about this.  ossl_quic_conn_poll_events is called from poll_translate_ssl_quic, which does this:

```
if (!ossl_rio_poll_builder_add_fd(rpb, fd_nfy, /*r=*/1, /*w=*/0))
            return 0;

        /* Tell QUIC domain we need to receive notifications. */
        ossl_quic_enter_blocking_section(ssl, wctx);

        /*
         * Only after the above call returns is it guaranteed that any readiness
         * events will cause the above notifier to become readable. Therefore,
         * it is possible the object became ready after our initial
         * poll_readout() call (before we determined that nothing was ready and
         * we needed to block). We now need to do another readout, in which case
         * blocking is to be aborted.
         */
        if (!ossl_quic_conn_poll_events(ssl, events, /*do_tick=*/0, &revents)) {
            ossl_quic_leave_blocking_section(ssl, wctx);
            return 0;
        }

        if (revents != 0) {
            ossl_quic_leave_blocking_section(ssl, wctx);
            *abort_blocking = 1;
            return 1;
        }
    }
```

The notable part there is the call to ossl_quic_enter_blocking_section.  From my read that function should only ever be called with the corresponding qctx_lock held, as its mutating data that other threads might be accessing as part of the reactor code.

While I don't see where it happens, if we assume a calling function to poll_translate_ssl_quic has already taken the qctx_lock, then this would cause a deadlock.  If its not (i.e. if we assume the converse, that the qctx_lock was not held previously), then we have a potential race condition on the call to enter_critical_section.

I think the 'right' thing to do is to move that lock to the calling functions poll_translate_ssl_quic and poll_readout

Reply:
I've added locks to ossl_quic_enter_blocking_section() and ossl_quic_leave_blocking_section() calls. That's IMO the right way to fix it.

Reply:
yeah, that works for me as a solution